src/main/java/com/exasol/errorreporting/ErrorMessageBuilder.java

```diff
@@ -0,0 +1,158 @@
+     */
+    public ErrorMessageBuilder parameter(final String placeholder, final Object value, final String description) {
+        return this.parameter(placeholder, value);
+    }
+
+    /**
+     * Add a parameter without quotes.
+     * 
+     * @param placeholder placeholder without parentheses
+     * @param value       value to insert
+     * @return self for fluent programming
+     */
+    public ErrorMessageBuilder unquotedParameter(final String placeholder, final Object value) {
+        this.parameterMapping.put(placeholder, value.toString());
+        return this;
+    }
+
+    /**
+     * Add a parameter without quotes.
+     *
+     * @param placeholder placeholder without parentheses
+     * @param value       value to insert
+     * @param description description for the error catalog
+     * @return self for fluent programming
+     */
+    public ErrorMessageBuilder unquotedParameter(final String placeholder, final Object value,
+            final String description) {
+        return this.unquotedParameter(placeholder, value);
+    }
+
+    /**
+     * Add a mitigation. Explain here what users can do to resolve or avoid this error.
+     *
+     * @param mitigation explanation
+     * @return self for fluent programming
+     */
+    public ErrorMessageBuilder mitigation(final String mitigation) {
+        this.mitigations.add(mitigation);
+        return this;
+    }
+
+    /**
+     * Build the error message.
+     * 
+     * @return built error massage string
+     */
+    public String toString() {
+        final StringBuilder result = new StringBuilder();
+        result.append(this.errorCode);
+        if (this.messageBuilder.length() > 0) {
+            result.append(": ");
+            result.append(applyPlaceholders(this.messageBuilder.toString()));
+        }
+        if (this.mitigations.size() == 1) {
+            result.append(" ");
+            result.append(this.mitigations.get(0));
+        } else if (this.mitigations.size() > 1) {
+            result.append(" Known mitigations:");
+            this.mitigations.stream().forEach(mitigation -> {
+                result.append("\n* ");
+                result.append(mitigation);
+            });
+        }
+        return result.toString();
+    }
+
+    private String applyPlaceholders(final String subject) {
+        return replacePlaceholders(subject, this::resolvePlaceholder);
+    }
+
+    private String resolvePlaceholder(final String placeholder) {
+        if (this.parameterMapping.containsKey(placeholder)) {
+            return this.parameterMapping.get(placeholder);
+        } else {
+            throw new IllegalStateException(ExaError.messageBuilder("F-ERJ-1")//
+                    .message("Unknown placeholder {{placeholder}}.").parameter("placeholder", placeholder)//
+                    .toString());
+        }
+    }
+
+    private String replacePlaceholders(final String subject, final UnaryOperator<String> placeholderResolver) {
```

Comment:
Why do we need the placeholderResolver parameter?

can't we invoke it directly from below?

Reply:
I did not want to chain methods. Using this method parameter I emulate a more convenient API (using callback) than the strange Java one.